core/trino-main/src/test/java/io/trino/operator/TestExchangeClient.java

```diff
@@ -391,6 +392,80 @@ public void testStreamingAddLocation()
         assertStatus(statuses.get(location2), location2, "closed", 3, 3, 3, "not scheduled");
     }
 
+    @Test
+    public void testDeduplication()
+            throws Exception
+    {
+        DataSize maxResponseSize = DataSize.of(10, Unit.MEGABYTE);
+        MockExchangeRequestProcessor processor = new MockExchangeRequestProcessor(maxResponseSize);
+
+        TaskId taskP0A0 = new TaskId(new StageId("query", 1), 0, 0);
+        TaskId taskP1A0 = new TaskId(new StageId("query", 1), 1, 0);
+        TaskId taskP0A1 = new TaskId(new StageId("query", 1), 0, 1);
+
+        URI locationP0A0 = URI.create("http://localhost:8080/1");
+        URI locationP1A0 = URI.create("http://localhost:8080/2");
+        URI locationP0A1 = URI.create("http://localhost:8080/3");
+
+        processor.addPage(locationP1A0, createSerializedPage("location-1-page-1"));
+        processor.addPage(locationP0A1, createSerializedPage("location-2-page-1"));
+        processor.addPage(locationP0A1, createSerializedPage("location-2-page-2"));
+
+        @SuppressWarnings("resource")
+        ExchangeClient exchangeClient = new ExchangeClient(
+                "localhost",
+                DataIntegrityVerification.ABORT,
+                new DeduplicationExchangeClientBuffer(scheduler, DataSize.of(1, Unit.KILOBYTE), RetryPolicy.QUERY),
+                maxResponseSize,
+                1,
+                new Duration(1, SECONDS),
+                true,
+                new TestingHttpClient(processor, scheduler),
+                scheduler,
+                new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext(), "test"),
+                pageBufferClientCallbackExecutor);
+
+        exchangeClient.addLocation(taskP0A0, locationP0A0);
+        exchangeClient.addLocation(taskP1A0, locationP1A0);
+        exchangeClient.addLocation(taskP0A1, locationP0A1);
+
+        processor.setComplete(locationP0A0);
+        processor.setFailed(locationP1A0, new RuntimeException("failure"));
```

Comment:
I do not get this test. Why we are getting results if for `taskP1A0` we get failure.

Reply:
To verify full query retry semantics. With full query retries the assumption is that all tasks are restarted, and the number of tasks may not match. In this tests the first attempt has 2 tasks and the second attempt has only one. Since the only task for the second attempt succeeded - it is considered a success.

Reply:
makes sense. Add the comment to test though.